src/CoreWCF.Kafka/src/CoreWCF/Channels/KafkaTransportPump.cs

```diff
@@ -0,0 +1,246 @@
+        new(@"^[a-zA-Z0-9\.\-_]{1,255}$", RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
+
+    public KafkaTransportPump(KafkaTransportBindingElement transportBindingElement,
+        ILogger<KafkaTransportPump> logger,
+        IServiceDispatcher serviceDispatcher, KafkaDeliverySemantics kafkaDeliverySemantics)
+    {
+        _logger = logger;
+        _kafkaDeliverySemantics = kafkaDeliverySemantics;
+        Topic = serviceDispatcher.BaseAddress.PathAndQuery.TrimStart('/');
+        if (string.IsNullOrEmpty(Topic) || !s_topicNameRegex.IsMatch(Topic))
+        {
+            throw new NotSupportedException(string.Format(SR.InvalidTopicName, Topic));
+        }
+
+        if (transportBindingElement.ErrorHandlingStrategy == KafkaErrorHandlingStrategy.DeadLetterQueue)
+        {
+            if (string.IsNullOrEmpty(transportBindingElement.DeadLetterQueueTopic))
+            {
+                throw new NotSupportedException(SR.InvalidDeadLetterQueueTopicName);
+            }
+
+            if (!s_topicNameRegex.IsMatch(transportBindingElement.DeadLetterQueueTopic))
+            {
+                throw new NotSupportedException(string.Format(SR.InvalidTopicName, Topic));
+            }
+        }
+        TransportBindingElement = transportBindingElement;
+        _baseAddress = serviceDispatcher.BaseAddress;
+    }
+
+    public override Task StartPumpAsync(QueueTransportContext queueTransportContext, CancellationToken token)
+    {
+        _cts = new();
+        _mres = new();
+        _mres.Reset();
+        _receiveContextCountdownEvent = new(1);
+
+        CancellationToken ct = CancellationTokenSource
+            .CreateLinkedTokenSource(_cts.Token, token).Token;
+
+        ConsumerConfig = TransportBindingElement.Config;
+        ConsumerConfig.BootstrapServers = _baseAddress.Authority;
+        var (enableAutoCommit, enableAutoOffsetStore) = GetCommitStrategyConfigValues(ConsumerConfig, _kafkaDeliverySemantics);
+        ConsumerConfig.EnableAutoCommit = enableAutoCommit;
+        ConsumerConfig.EnableAutoOffsetStore = enableAutoOffsetStore;
+        Consumer = new ConsumerBuilder<Null, byte[]>(ConsumerConfig)
+            .SetKeyDeserializer(Deserializers.Null)
+            .SetValueDeserializer(Deserializers.ByteArray)
+            .SetLogHandler(OnLog)
+            .SetErrorHandler(OnError)
+            .Build();
+
+        Consumer.Subscribe(Topic);
+
+        if (TransportBindingElement.ErrorHandlingStrategy == KafkaErrorHandlingStrategy.DeadLetterQueue)
+        {
+            ProducerConfig producerConfig = new(ConsumerConfig);
+            producerConfig.Acks = Acks.All;
+            Producer = new ProducerBuilder<Null, byte[]>(producerConfig)
+                .SetKeySerializer(Serializers.Null)
+                .SetValueSerializer(Serializers.ByteArray)
+                .Build();
+        }
+
+        Task.Run(async () =>
+        {
+            while (!ct.IsCancellationRequested)
+            {
```

Comment:
The lifetime validity of the CancellationToken passed in to StartPunpAsync is until the Task returned from StartPumpAsync completes. Here you are creating a background task using Task.Run which captures ct in the delegate, and ct is a token linked to the token passed to StartPumpAsync. After the Task returned from the method StartPumpAsync completes, cancelling the CancellationToken shouldn't have side effects, but it does.  
I don't think you need to link them. If you do link them, you should keep a reference to the CancellationTokenSource returned from calling CreateLinkedTokenSource as you need to Dispose it when you're done. Otherwise you can get a memory leak where it keeps itself alive. A bit like you can get when you add an EventHandler to an event and you don't remove it.

Reply:
done